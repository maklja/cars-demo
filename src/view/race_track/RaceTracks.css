.race-tracks-overview {

	/* class that provides layout for all tracks */
	text-align: center;
	margin-bottom: 6em;
}

.race-tracks-overview .race-tracks-overview__title {
	font-size: 3em;
	margin: 1em 0;
}

.race-tracks-overview .race-tracks-overview__message {

	/* Styles for message when no cars are selected */
	font-size: 2em;
	padding: 6em 0;
	background-color: #eeeeee;
}

.race-tracks-overview .race-tracks-overview__button {

	/* start button styles */
	text-align: right;
}

.race-track {

	/* Race track styles */
	background-color: #eeeeee;
	margin: 0.6em 0;
	border: 0.063em solid #cccccc;
	border-radius: 1em;
	position: relative;
	min-height: 10em;
	overflow: hidden;
}

.race-track .race-track__wrapper {
	position: absolute;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
	z-index: 2; /* put image on top track line and finish line */
	transition: transform 0s;
}

.race-track .race-track__wrapper .race-track__finish {
	display: flex;
	align-items: center;
	border-radius: 1em;
	width: 12em;
	height: 100%;
	transition: transform 0s; /* for this part of task just put image at the finish line */
}

.race-track .race-track__finish.race-track__finish--first-place {
	background-color: rgba(255, 223, 0, 0.8);
}

.race-track .race-track__finish.race-track__finish--second-place {
	background-color: rgba(192, 192, 192, 0.8);
}

.race-track .race-track__finish.race-track__finish--third-place {
	background-color: rgba(205, 127, 50, 0.8);
}

.race-track .race-track__wrapper .race-track__place-number {
	position: absolute;
	top: 0;
	left: 0;
	width: 100%;
	height: 100%;
	display: flex;
	justify-content: center;
	align-items: center;
	font-size: 6em;
	color: white;
}

.race-track .race-track__wrapper .race-track__image {

	/* car image on the track */
	max-width: 100%;
	width: 100%;
}

.race-track.race-track--race-end .race-track__wrapper {
	transform: translateX(100%);
}

.race-track.race-track--race-end .race-track__wrapper .race-track__finish {
	transform: translateX(-100%);
}

.race-track .race-track__line {

	/* Track lines on the middle */
	border: 0.063em dashed black;
	position: absolute;
	top: 50%;
	left: 1em;
	right: 1em;
}

.race-track .race-track__finish-line {

	/* Finish line styles made from png image */
	height: 100%;
	position: absolute;
	top: 0;
	right: 0;
	width: 2em;
	background-image: url("../../assets/images/finish_line.png");
	background-size: 200%;
}
