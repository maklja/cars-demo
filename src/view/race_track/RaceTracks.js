import PropTypes from 'prop-types';
import React, { Component } from 'react';

import { carPropTypes } from '../../props';

import './RaceTracks.css';

const medalColorsClasses = [
	'race-track__finish--first-place',
	'race-track__finish--second-place',
	'race-track__finish--third-place'
];

export const RaceTrack = ({ name, image, speed, isRaceStarted, place }) => {
	const placeNumber =
		// check range of car place and create view if needed
		0 <= place && place < medalColorsClasses.length ? (
			<div className="race-track__place-number">
				<div>{place + 1}</div>
			</div>
		) : null;

	return (
		<div
			className={`race-track ${
				isRaceStarted ? 'race-track--race-end' : ''
			}`}
			title={`Top speed: ${speed}`}
		>
			<div className="race-track__wrapper">
				<div
					className={`race-track__finish ${
						medalColorsClasses[place]
					}`}
				>
					<img className="race-track__image" src={image} alt={name} />
					{placeNumber}
				</div>
			</div>
			<div className="race-track__line" />
			<div className="race-track__finish-line" />
		</div>
	);
};

RaceTrack.propTypes = {
	...carPropTypes,
	isRaceStarted: PropTypes.bool,
	place: PropTypes.number // car place at the end of the race
};

RaceTrack.defaultProps = {
	isRaceStarted: false,
	place: -1
};

class RaceTracks extends Component {
	constructor(props) {
		super(props);

		this.state = {
			isRaceStarted: false,
			positions: []
		};

		this._startRace = this._startRace.bind(this);
		this._resetRace = this._resetRace.bind(this);
	}

	render() {
		const { cars } = this.props;
		const { isRaceStarted, positions } = this.state;

		return (
			<div className="container-fluid race-tracks-overview">
				<div className="row">
					<div className="col-md-12">
						<div className="race-tracks-overview__title">
							Race tracks
						</div>
					</div>
					{cars.length > 0 ? (
						cars.map(curCar => (
							<div className="col-md-12" key={curCar.id}>
								<RaceTrack
									{...curCar}
									place={positions.indexOf(curCar.id)}
									isRaceStarted={isRaceStarted}
								/>
							</div>
						))
					) : (
						<div className="col-md-12">
							<div className="race-tracks-overview__message">
								Please select cars to start a race.
							</div>
						</div>
					)}
					<div className="col-md-12">
						<div className="race-tracks-overview__button">
							<button
								name="start-race"
								onClick={this._startRace}
								disabled={cars.length === 0 || isRaceStarted}
								className="button-control button-control--large"
							>
								Start
							</button>

							<button
								name="reset-race"
								onClick={this._resetRace}
								disabled={isRaceStarted === false}
								className="button-control button-control--large"
							>
								Reset
							</button>
						</div>
					</div>
				</div>
			</div>
		);
	}

	_startRace() {
		const { cars } = this.props;
		this.setState({
			isRaceStarted: true, // start the race
			// find out which car is fastest by speed
			positions: cars
				.concat() // don't mutate array
				.sort((carA, carB) => carB.speed - carA.speed) // sort cars by speed
				.map(curCar => curCar.id) // just take car id
		});
	}

	_resetRace() {
		// reset race
		this.setState({
			isRaceStarted: false,
			positions: []
		});
	}
}

RaceTracks.propTypes = {
	cars: PropTypes.arrayOf(PropTypes.shape(carPropTypes)).isRequired
};

export default RaceTracks;
