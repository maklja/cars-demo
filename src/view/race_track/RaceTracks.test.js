import React from 'react';
import renderer from 'react-test-renderer';
import { shallow, configure } from 'enzyme'; // use enzyme to simulate user interaction
import Adapter from 'enzyme-adapter-react-16';

import RaceTracks, { RaceTrack } from './RaceTracks';

// configure enzyme for react 16
configure({ adapter: new Adapter() });

describe('RaceTracks test', () => {
	it('RaceTracks renders correctly with no cars on track', () => {
		const cars = [];

		const tree = renderer.create(<RaceTracks cars={cars} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('RaceTracks renders correctly with three cars on track', () => {
		const cars = [
			{
				image:
					'https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png',
				speed: 235,
				description:
					'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
				name: 'Mazda 3',
				id: 2
			},
			{
				image: 'http://www.nriol.com/images/benz-cla.png',
				speed: 280,
				description:
					'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
				name: 'Mercedes Benz CLA',
				id: 3
			},
			{
				description:
					'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
				id: 1,
				image:
					'https://lb.assets.fiat.com/assets/00/punto-my16/images/thumbnail/equipment/street.png',
				name: 'Fiat Punto',
				speed: 148
			}
		];

		const tree = renderer.create(<RaceTracks cars={cars} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('check race result after start button is clicked', () => {
		const cars = [
			{
				image:
					'https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png',
				speed: 235,
				description:
					'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
				name: 'Mazda 3',
				id: 2
			},
			{
				image: 'http://www.nriol.com/images/benz-cla.png',
				speed: 280,
				description:
					'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
				name: 'Mercedes Benz CLA',
				id: 3
			},
			{
				description:
					'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
				id: 1,
				image:
					'https://lb.assets.fiat.com/assets/00/punto-my16/images/thumbnail/equipment/street.png',
				name: 'Fiat Punto',
				speed: 148
			},
			{
				description: 'Yugo',
				id: 11,
				image:
					'https://lb.assets.fiat.com/assets/00/punto-my16/images/thumbnail/equipment/street.png',
				name: 'Yugo',
				speed: 99
			}
		];

		const wrapper = shallow(<RaceTracks cars={cars} />);

		// find and click on start button
		wrapper.find('[name="start-race"]').simulate('click');

		// after race is over get positions that are sorted from fastest
		// to slowest car
		const placesAtRaceEnd = wrapper.state().positions;

		expect(placesAtRaceEnd).toEqual([
			3, // Mercedes Benz CLA has the first place
			2, // Mazda 3 has the second place
			1, // Fiat Punto has third place
			11 // Yugo has fourth place
		]);
	});
});

describe('RaceTrack tests', () => {
	it('RaceTrack renders correctly with Mazda 3 car data', () => {
		const car = {
			image:
				'https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png',
			speed: 235,
			description:
				'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
			name: 'Mazda 3',
			id: 2
		};

		const tree = renderer.create(<RaceTrack {...car} />).toJSON();
		expect(tree).toMatchSnapshot();
	});

	it('RaceTrack renders correctly with gold, silver, bronze place and fourth place', () => {
		const car = {
			image:
				'https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png',
			speed: 235,
			description:
				'The Fiat Punto is a supermini car produced by the Italian manufacturer Fiat since 1993.',
			name: 'Mazda 3',
			id: 2
		};

		const firstPlace = 0; // gold place
		const firstPlaceRaceTrack = renderer
			.create(<RaceTrack {...car} place={firstPlace} />)
			.toJSON();
		expect(firstPlaceRaceTrack).toMatchSnapshot();

		const secondPlace = 1; // silver place
		const secondPlaceRaceTrack = renderer
			.create(<RaceTrack {...car} place={secondPlace} />)
			.toJSON();
		expect(secondPlaceRaceTrack).toMatchSnapshot();

		const thirdPlace = 2; // bronze place
		const thirdPlaceRaceTrack = renderer
			.create(<RaceTrack {...car} place={thirdPlace} />)
			.toJSON();
		expect(thirdPlaceRaceTrack).toMatchSnapshot();

		const fourthPlace = 4; // no reward place
		const fourthPlaceRaceTrack = renderer
			.create(<RaceTrack {...car} place={fourthPlace} />)
			.toJSON();
		expect(fourthPlaceRaceTrack).toMatchSnapshot();
	});
});
