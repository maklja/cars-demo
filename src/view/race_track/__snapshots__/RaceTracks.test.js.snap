// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RaceTrack tests RaceTrack renders correctly with Mazda 3 car data 1`] = `
<div
  className="race-track "
  title="Top speed: 235"
>
  <div
    className="race-track__wrapper"
  >
    <div
      className="race-track__finish undefined"
    >
      <img
        alt="Mazda 3"
        className="race-track__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
    </div>
  </div>
  <div
    className="race-track__line"
  />
  <div
    className="race-track__finish-line"
  />
</div>
`;

exports[`RaceTrack tests RaceTrack renders correctly with gold, silver, bronze place and fourth place 1`] = `
<div
  className="race-track "
  title="Top speed: 235"
>
  <div
    className="race-track__wrapper"
  >
    <div
      className="race-track__finish race-track__finish--first-place"
    >
      <img
        alt="Mazda 3"
        className="race-track__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
      <div
        className="race-track__place-number"
      >
        <div>
          1
        </div>
      </div>
    </div>
  </div>
  <div
    className="race-track__line"
  />
  <div
    className="race-track__finish-line"
  />
</div>
`;

exports[`RaceTrack tests RaceTrack renders correctly with gold, silver, bronze place and fourth place 2`] = `
<div
  className="race-track "
  title="Top speed: 235"
>
  <div
    className="race-track__wrapper"
  >
    <div
      className="race-track__finish race-track__finish--second-place"
    >
      <img
        alt="Mazda 3"
        className="race-track__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
      <div
        className="race-track__place-number"
      >
        <div>
          2
        </div>
      </div>
    </div>
  </div>
  <div
    className="race-track__line"
  />
  <div
    className="race-track__finish-line"
  />
</div>
`;

exports[`RaceTrack tests RaceTrack renders correctly with gold, silver, bronze place and fourth place 3`] = `
<div
  className="race-track "
  title="Top speed: 235"
>
  <div
    className="race-track__wrapper"
  >
    <div
      className="race-track__finish race-track__finish--third-place"
    >
      <img
        alt="Mazda 3"
        className="race-track__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
      <div
        className="race-track__place-number"
      >
        <div>
          3
        </div>
      </div>
    </div>
  </div>
  <div
    className="race-track__line"
  />
  <div
    className="race-track__finish-line"
  />
</div>
`;

exports[`RaceTrack tests RaceTrack renders correctly with gold, silver, bronze place and fourth place 4`] = `
<div
  className="race-track "
  title="Top speed: 235"
>
  <div
    className="race-track__wrapper"
  >
    <div
      className="race-track__finish "
    >
      <img
        alt="Mazda 3"
        className="race-track__image"
        src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
      />
    </div>
  </div>
  <div
    className="race-track__line"
  />
  <div
    className="race-track__finish-line"
  />
</div>
`;

exports[`RaceTracks test RaceTracks renders correctly with no cars on track 1`] = `
<div
  className="container-fluid race-tracks-overview"
>
  <div
    className="row"
  >
    <div
      className="col-md-12"
    >
      <div
        className="race-tracks-overview__title"
      >
        Race tracks
      </div>
    </div>
    <div
      className="col-md-12"
    >
      <div
        className="race-tracks-overview__message"
      >
        Please select cars to start a race.
      </div>
    </div>
    <div
      className="col-md-12"
    >
      <div
        className="race-tracks-overview__button"
      >
        <button
          className="button-control button-control--large"
          disabled={true}
          name="start-race"
          onClick={[Function]}
        >
          Start
        </button>
        <button
          className="button-control button-control--large"
          disabled={true}
          name="reset-race"
          onClick={[Function]}
        >
          Reset
        </button>
      </div>
    </div>
  </div>
</div>
`;

exports[`RaceTracks test RaceTracks renders correctly with three cars on track 1`] = `
<div
  className="container-fluid race-tracks-overview"
>
  <div
    className="row"
  >
    <div
      className="col-md-12"
    >
      <div
        className="race-tracks-overview__title"
      >
        Race tracks
      </div>
    </div>
    <div
      className="col-md-12"
    >
      <div
        className="race-track "
        title="Top speed: 235"
      >
        <div
          className="race-track__wrapper"
        >
          <div
            className="race-track__finish undefined"
          >
            <img
              alt="Mazda 3"
              className="race-track__image"
              src="https://www.mazdausa.com/siteassets/vehicles/2017/m3s/trims/sport/17-m3s-sp-25d-snowflake-white.png"
            />
          </div>
        </div>
        <div
          className="race-track__line"
        />
        <div
          className="race-track__finish-line"
        />
      </div>
    </div>
    <div
      className="col-md-12"
    >
      <div
        className="race-track "
        title="Top speed: 280"
      >
        <div
          className="race-track__wrapper"
        >
          <div
            className="race-track__finish undefined"
          >
            <img
              alt="Mercedes Benz CLA"
              className="race-track__image"
              src="http://www.nriol.com/images/benz-cla.png"
            />
          </div>
        </div>
        <div
          className="race-track__line"
        />
        <div
          className="race-track__finish-line"
        />
      </div>
    </div>
    <div
      className="col-md-12"
    >
      <div
        className="race-track "
        title="Top speed: 148"
      >
        <div
          className="race-track__wrapper"
        >
          <div
            className="race-track__finish undefined"
          >
            <img
              alt="Fiat Punto"
              className="race-track__image"
              src="https://lb.assets.fiat.com/assets/00/punto-my16/images/thumbnail/equipment/street.png"
            />
          </div>
        </div>
        <div
          className="race-track__line"
        />
        <div
          className="race-track__finish-line"
        />
      </div>
    </div>
    <div
      className="col-md-12"
    >
      <div
        className="race-tracks-overview__button"
      >
        <button
          className="button-control button-control--large"
          disabled={false}
          name="start-race"
          onClick={[Function]}
        >
          Start
        </button>
        <button
          className="button-control button-control--large"
          disabled={true}
          name="reset-race"
          onClick={[Function]}
        >
          Reset
        </button>
      </div>
    </div>
  </div>
</div>
`;
